### DISCUSSION TITLE
What features would you like to see here?

### DISCUSSION THREAD

**Discussion by SteveDunn** [SteveDunn/Intellenum#11] [Q&A]
The main idea behind Intellenum is for a source generator to generate enums and the associated `Contains`, `GetValues`. The other goal is for speed and memory allocations to closely match that of using enums directly.

What features would you like to see?

What features would you *not* like to see? e.g. design decisions in similar libraries that haven't turned out so well.

Here's what's already there (and subject to change depending on this discussion)

* ability to create `class` or `struct` enums - if memory / GC is important, then this will be useful
* very similar analyzers - you won't be able to accidentally create additional enum values outside of the declaring type
* able to create instances with attributes, e.g. `[Instance("Male", "M")]`

Here's what I'm thinking might be useful:

* ability to create instances like
```csharp
[Intellenum("Tiger, Leopard, Tortoise")
public partial struct Animal { }

Console.WriteLine(Animal.Tiger); => "Tiger"
Console.WriteLine(Animal.Tiger.StringIdentifier); => "Tiger"
Console.WriteLine(Animal.Tiger.NumbericIdentifier); => "0"
```

* ability to quickly see if an enum contains a value
Code will be generated, something like:
```csharp
public static bool Contains(string value) {
  if(value == "Tiger") return true;
  if(value == "Leopard") return true;
  if(value == "Tortoise") return true;
  return false;
}
```

* ability to quickly iterate
```csharp
public static IEnumerable<Animal> GetItems() {
  yield return Tiger;
  yield return Leopard;
  yield return Tortoise;
}
```

**Comment by viceroypenguin** [comment]
FYI only. Something else in this space, though very different: [domn1995/dunet](https://github.com/domn1995/dunet). I don't see this as overlapping that (pure enum vs DU), but there's always something to be learned from similar libraries in the space.

**Reply by SteveDunn** [reply]
Interesting library - thanks for the link. I might consider that one over 'OneOf' that I'm currently using

**Comment by JoshuaNitschke** [comment]
Hi Steve, I'm really enjoying Vogen and decided to give this lib a try too. One thing I noticed is that I have to updated all my 

`if(enum is Enum.Member)` to `if(enum == Enum.Member)`

It would be nice if we can use the `is` syntax

**Reply by SteveDunn** [reply]
It'd be nice to have this, but I think it only works with const values
